### PR TITLE
Add clang-format

### PR DESCRIPTION
**Pull request by gergondet**
This PR adds a clang-format file to the repository and apply it to the code. I've added some formatting blocks where I could identify it was useful (I've mostly checked the style effect on examples and public includes)

While at it, I've updated all copyright notices to use the short form.

I'm opening this as a PR so we can have a look at the style (and maybe change it) before we merge it.  I advise it's looked at with whitespace changes filtered out otherwise it's unreadable :D

### CODE REVIEW

**Review by aescande — COMMENTED** [review]
Overall, I'm fine with the changes.

One thing I'm puzzled with is the decision to pack/unpack one-liner. I don't really like some of the unpacking especially in TMP situtation (CompiledAssignment.h is a concentrate of the problem), namely:
 - empty class and struct get expanded,
 - it seems like all templated one-liners get expanded.

Other than that, a not too important change could be to not remove spaces before C++-style comments to keep comments aligned.

**Comment by aescande** [comment]
Oh, another point: I'd rather have the order of include disregard the upper and lower case. Otherwise it leads to some strange situations where e.g `tvm/defs.h` get included between `tvm/constraint/long/path` and `tvm/function/another/path`.

File: include/tvm/solver/LSSOLLeastSquareSolver.h [hunk 1]
```diff
@@ -4,107 +4,105 @@
 
 #include <tvm/solver/abstract/LeastSquareSolver.h>
 
+#include <eigen-lssol //LSSOL_FP.h>
```
[the remark this answers is not in the document]
Comment (gergondet) [reply include/tvm/solver/LSSOLLeastSquareSolver.h]: Yup, but this was in the original :)

Fixed in 2c092000926833c4573b15a371647d3d9fc0f485

File: tests/doctest/doctest.h [hunk 2]
```diff
@@ -56,8 +56,7 @@
 #define DOCTEST_VERSION_PATCH 6
 #define DOCTEST_VERSION_STR "1.2.6"
 
-#define DOCTEST_VERSION                                                                            \
-    (DOCTEST_VERSION_MAJOR * 10000 + DOCTEST_VERSION_MINOR * 100 + DOCTEST_VERSION_PATCH)
+#define DOCTEST_VERSION (DOCTEST_VERSION_MAJOR * 10000 + DOCTEST_VERSION_MINOR * 100 + DOCTEST_VERSION_PATCH)
```
[the remark this answers is not in the document]
Comment (gergondet) [reply tests/doctest/doctest.h]: Nope, I forgot it was there, reverted in 91e9252e

File: examples/LeastSquaresExample.cpp [hunk 3]
```diff
@@ -60,11 +68,11 @@ bool leastSquares3points()
...
-  Vector3d p1(-1, -1, 0);                     // Contact point position
-  Vector3d n1(1, 1, 1);                       // Normal vector to contact surface
-  MatrixXd R1 = rotationFromZ(n1);            // Rotation between world frame and a local contact frame
-  Vector3d f1des = Vector3d::Zero();          // Desired force for f1
-  VariablePtr f1 = S.createVariable("f1");    // Variable creation
+  Vector3d p1(-1, -1, 0); // Contact point position
+  Vector3d n1(1, 1, 1); // Normal vector to contact surface
+  MatrixXd R1 = rotationFromZ(n1); // Rotation between world frame and a local contact frame
+  Vector3d f1des = Vector3d::Zero(); // Desired force for f1
+  VariablePtr f1 = S.createVariable("f1"); // Variable creation
```
[the remark this answers is not in the document]
Comment (gergondet) [reply examples/LeastSquaresExample.cpp:75]: Changed comment alignment styles in 31fc3bc9c0eed87d2c2ef6f773b12acff642ac25

File: include/tvm/utils/ProtoTask.h [hunk 4]
```diff
@@ -39,195 +39,286 @@ namespace tvm
...
+ * not 0), otherwise the compiler won't be able to decide wich overload to
+ * pick between this and shared_ptr operator.
+ */
 /**@{*/
 template<typename F>
-inline tvm::utils::ProtoTaskEQRet<F> operator==(std::shared_ptr<F> f, const tvm::utils::internal::RHS& rhs) { return { f, rhs }; }
+inline tvm::utils::ProtoTaskEQRet<F> operator==(std::shared_ptr<F> f, const tvm::utils::internal::RHS & rhs)
+{
+  return {f, rhs};
+}
```
[the remark this answers is not in the document]
Comment (gergondet) [reply include/tvm/utils/ProtoTask.h]: Reverted in ed7260af

**Comment by gergondet** [comment]
> One thing I'm puzzled with is the decision to pack/unpack one-liner. I don't really like some of the unpacking especially in TMP situtation (CompiledAssignment.h is a concentrate of the problem), namely:

I'm not a huge fan too and I've spent a good chunk of my morning on fiddling with clang-format options and I came up with a few options (and discovered that for some reasons templated functions are not treated).

Here's what I can get: https://gist.github.com/gergondet/004cbde7aec9b27cb8aab34fbbff24c2
- option 1: `BreakBeforeBraces: Allman` This is the style applied in the MR
- option 2: `BreakBeforeBraces: Custom` (same break rules as Allman but SplitEmpty* is false) Very close to the style applied to this MR and waste a little less space
- option 3: `BreakBeforeBraces: Custom` (no break after class/struct/function) gets a lot of short stuff into a single line but a big departure from our more usual style
- option 4: `BreakBeforeBraces:Custom` (same as option 3) plus `AlwaysBreakTemplateDeclarations: MultiLine` gets everything that fits on one line in one line similar to option 3 when the template declaration gets too long to have everything fit one line

Note: the main goal of clang-format is to have a blanket solution for 99% of the code and `// clang-format off`/`// clang-format on` can come to the rescue when the formatting goes against the tool but makes sense locally

> Oh, another point: I'd rather have the order of include disregard the upper and lower case. Otherwise it leads to some strange situations where e.g tvm/defs.h gets included between tvm/constraint/long/path and tvm/function/another/path.

This one has to do with sorting directly and not really lower/upper case. Two ways to do that:
- manually split includes into groups so that `tvm/defs.h` is first
- fiddle with `IncludeBlocks: Regroup` and `IncludeCategories` from my experience this is difficult to get right

**Comment by aescande** [comment]
Thank you for the MR 👍 . I'll merge once the CI has finished.

**Pull request merged**